server/build_event_protocol/build_event_handler/build_event_handler.go

```diff
@@ -883,18 +898,32 @@ func (e *EventChannel) collectAPIFacets(iid string, event *build_event_stream.Bu
 		// early exit if this isn't an action event.
 		return nil
 	}
-	b, err := proto.Marshal(action)
+	buf, err := proto.Marshal(action)
 	if err != nil {
 		return err
 	}
 	key := api_common.ActionsKey(iid)
-	if err := e.collector.ListAppend(e.ctx, key, string(b)); err != nil {
-		return err
+	if _, ok := e.compressedMetricWriters[key]; !ok {
+		wf := writerFunc(func(data []byte) (int, error) {
+			if err := e.collector.ListAppend(e.ctx, key, string(data)); err != nil {
+				return 0, err
+			}
+			if err := e.collector.Expire(e.ctx, key, apiFacetsExpiration); err != nil {
+				return 0, err
+			}
+			return len(data), nil
+		})
+		w, err := zstd.NewWriter(wf)
```

Comment:
If each of these writers processes only a small amount of data (less than 1MB or so) then it might be worth benchmarking this vs. chunk-by-chunk compression (e.g. `ListAppend(..., compression.CompressZstd(data))`) -- I remember experiencing some issues with the NewWriter API for small chunks, and getting significantly better performance by having the NewZstdChunkingCompressor reuse a single Writer instance.